### PR TITLE
Encapsulate editor/editor_mode

### PR DESCRIPTION
**Pull request by neiljp**
This was previously partially encapsulated in WriteBox for local use,
but this wraps all functionality with access via the Controller using:
* is_in_editor_mode()
* enter_edit_mode_with(editor)
* exit_editor_mode()
* current_editor()

This then allows further tests for this and refactoring, as well as being more descriptive.

### CODE REVIEW

File: tests/ui_tools/test_boxes.py [hunk 1]
```diff
@@ -208,7 +208,9 @@ def test_reset_search_text(self, panel_search_box):
     def test_keypress_ENTER(self, panel_search_box,
                             enter_key, log, expect_body_focus_set):
         size = (20,)
-        panel_search_box.panel_view.view.controller.editor_mode = True
+        panel_search_box.panel_view.view.controller.is_in_editor_mode = (
```
Comment (sumanthvrao) [inline tests/ui_tools/test_boxes.py:211]: The previous diff looked like it was setting the editor mode to True, which according to our helper functions should have called `enter_edit_mode_with()`. Did you use `is_in_editor_mode`  to makes things simpler? If that is the case, maybe adding a comment here would help.

File: tests/ui/test_ui.py [hunk 2]
```diff
@@ -249,29 +253,31 @@ def test_keypress_autohide_streams(self, view, mocker, autohide, key):
...
-        assert view.controller.editor == view.stream_w.stream_search_box
+        (view.controller.enter_editor_mode_with
+         .assert_called_once_with(view.stream_w.stream_search_box))
 
     @pytest.mark.parametrize('key', keys_for_command('SEARCH_PEOPLE'))
     def test_keypress_edit_mode(self, view, mocker, key):
         view.users_view = mocker.Mock()
         view.body = mocker.Mock()
         view.user_search = mocker.Mock()
-        size = (20,)
```
Comment (sumanthvrao) [inline tests/ui/test_ui.py:269]: Move this dead code to a separate commit?

File: tests/core/test_core.py [hunk 3]
```diff
@@ -43,6 +43,23 @@ def test_initialize_controller(self, controller, mocker) -> None:
         self.model.poll_for_events.assert_called_once_with()
         assert controller.theme == self.theme
 
+    def test_initial_editor_mode(self, controller):
```
Comment (sumanthvrao) [inline tests/core/test_core.py:46]: Maybe extract this change along with the core.py changes into a previous commit?

File: zulipterminal/core.py [hunk 4]
```diff
@@ -34,8 +34,7 @@ def __init__(self, config_file: str, theme: ThemeSpec,
         self.autohide = autohide
         self.notify_enabled = notify
 
-        self._editor_mode = False
-        self._editor = None  # type: Any
+        self._editor = None  # type: Optional[Any]
```
Comment (sumanthvrao) [inline zulipterminal/core.py:37]: If I understand this right, we are doing away with `self._editor_mode` altogether. Maybe, add a small note in commit description about this?

File: tests/ui_tools/test_boxes.py (hunk 1 above)
Comment (neiljp) [reply tests/ui_tools/test_boxes.py:211]: What do you mean by 'previous diff'? Do you mean what this commit replaces? I agree that setting it to `True` as it stands (without encapsulation) is literally the same as setting editor_mode to True via doing `enter_edit_mode_with`. However, that assumes we have all of the state in the controller available. Instead of mocking the controller state (as previously), these lambda methods are essentially mocking the controller behavior (on the assumption that it doesn't matter that `is_in_editor_mode` *always* returns `True` in this test).

File: tests/core/test_core.py (hunk 3 above)
Comment (neiljp) [reply tests/core/test_core.py:46]: These tests are for the new functions, which replace the previous direct setting of the attributes. Do you mean to add the new methods and tests in one commit, and then migrate the code to use them in another? I suppose I could make the attributes private in the second step.

File: tests/ui/test_ui.py (hunk 2 above)
Comment (neiljp) [reply tests/ui/test_ui.py:269]: I think there's just this one line and a comment? That seems rather minor to have separately?

File: tests/ui_tools/test_boxes.py (hunk 1 above)
Comment (sumanthvrao) [reply tests/ui_tools/test_boxes.py:211]: Re `previous diff`, I meant what this commit replaces. Yup. Thanks for clarifying. Re state: That was what I suspected as well. :)

File: tests/core/test_core.py (hunk 3 above)
Comment (sumanthvrao) [reply tests/core/test_core.py:46]: Yes, exactly.

**Comment by neiljp** [comment]
@sumanthvrao Thanks for the review :+1:
